File: src/main/kotlin/MapHelper.java
import model.*;

import java.util.*;
import java.util.stream.Collectors;

import static java.lang.StrictMath.*;

public class MapHelper {

    public static final List<MapLine> mapLines = new ArrayList<>();
    public static final List<Point2D> mapPoints = new ArrayList<>();

    public static double mapSize = 4000D;

    public static final Point2D friendBasePoint = new Point2D(100.0D, mapSize - 100.0D);
    public static Point2D topPoint = new Point2D(300.0D, 300.0D);
    public static Point2D middlePoint = new Point2D(1659.0D, 1887);
    public static Point2D bottomPoint = new Point2D(mapSize - 100.0D, mapSize - 100.0D);
    public static Point2D enemyBasePoint = new Point2D(mapSize - 100.0D, 100.0D);

    public static MapLine topFriendLine = new MapLine(friendBasePoint, topPoint, LaneType.TOP, false);
    public static MapLine topEnemyLine = new MapLine(topPoint, enemyBasePoint, LaneType.TOP, true);

    public static MapLine middleFriendLine = new MapLine(friendBasePoint, middlePoint, LaneType.MIDDLE, false);
    public static MapLine middleEnemyLine = new MapLine(middlePoint, enemyBasePoint, LaneType.MIDDLE, true);

    public static MapLine bottomFriendLine = new MapLine(friendBasePoint, bottomPoint, LaneType.BOTTOM, false);
    public static MapLine bottomEnemyLine = new MapLine(bottomPoint, enemyBasePoint, LaneType.BOTTOM, true);

    public static MapLine artifactTopLine = new MapLine(topPoint, middlePoint, null);
    public static MapLine artifactBottomLine = new MapLine(middlePoint, bottomPoint, null);

    static {
        topFriendLine.getStartLines().addAll(Arrays.asList(middleFriendLine, bottomFriendLine));
        topFriendLine.getEndLines().addAll(Arrays.asList(topEnemyLine, artifactTopLine));

        topEnemyLine.getStartLines().addAll(Arrays.asList(topFriendLine, artifactTopLine));
        topEnemyLine.getEndLines().addAll(Arrays.asList(middleEnemyLine, bottomEnemyLine));

        middleFriendLine.getStartLines().addAll(Arrays.asList(topFriendLine, bottomFriendLine));
        middleFriendLine.getEndLines().addAll(Arrays.asList(topEnemyLine, artifactBottomLine, artifactTopLine));

        middleEnemyLine.getStartLines().addAll(Arrays.asList(topFriendLine, artifactBottomLine, artifactTopLine));
        middleEnemyLine.getEndLines().addAll(Arrays.asList(topEnemyLine, bottomEnemyLine));

        bottomFriendLine.getStartLines().addAll(Arrays.asList(topFriendLine, middleFriendLine));
        bottomFriendLine.getEndLines().addAll(Arrays.asList(artifactBottomLine, bottomEnemyLine));

        bottomEnemyLine.getStartLines().addAll(Arrays.asList(artifactBottomLine, bottomFriendLine));
        bottomEnemyLine.getEndLines().addAll(Arrays.asList(topEnemyLine, middleEnemyLine));

        artifactTopLine.getStartLines().addAll(Arrays.asList(topFriendLine, topEnemyLine));
        artifactTopLine.getEndLines().addAll(Arrays.asList(middleFriendLine, middleEnemyLine, artifactBottomLine));

        artifactBottomLine.getStartLines().addAll(Arrays.asList(middleFriendLine, middleEnemyLine, artifactTopLine));
        artifactBottomLine.getEndLines().addAll(Arrays.asList(bottomFriendLine, bottomEnemyLine));

        mapLines.addAll(Arrays.asList(
                topFriendLine, topEnemyLine, middleFriendLine, middleEnemyLine, bottomFriendLine, bottomEnemyLine, artifactTopLine, artifactBottomLine
        ));

        mapPoints.addAll(Arrays.asList(
                friendBasePoint, topPoint, middlePoint, bottomPoint, enemyBasePoint
        ));
    }

    public static final double LINE_RESOLVING_POSITION = 200D;
    public static final double LINE_RESOLVING_DISTANCE = 200D;

    public static final double DEAD_TOWER_HP_FACTOR = 0.1D;

    public static Map<Point2D, Building> deadGuardTowers = new HashMap<>();

    public Wizard wizard;
    public FindHelper findHelper;

    public MapHelper(World world, Game game, Wizard wizard) {
        this.wizard = wizard;
        this.findHelper = new FindHelper(world, game, wizard);

        updateLineInfo();
    }

    private void updateLineInfo() {
        clearLinesInfo();

        updateTowerInfo();
        updateWizardPositions();
        updateMinionPositions();
        updateStatuses();
    }

    private void clearLinesInfo() {
        mapLines.forEach(mapLine -> {
            mapLine.getEnemyWizardPositions().clear();
            mapLine.getFriendWizardPositions().clear();

            if (mapLine.getEnemy()) {
                mapLine.setEnemyPosition(0);
                mapLine.setFriendPosition(0);
            } else {
                mapLine.setEnemyPosition(mapLine.getLineLength());
                mapLine.setFriendPosition(mapLine.getLineLength());
            }

            mapLine.setDeadEnemyTowerCount(0);
            mapLine.setDeadFriendTowerCount(0);
        });
    }

    private void updateStatuses() {
        mapLines.forEach(line -> {
            if (line.getEnemyPosition() > 0) {
                line.setMapLineStatus(MapLineStatus.YELLOW);
                return;
            }

            if (line.getEnemyWizardPositions().size() > 0) {
                line.setMapLineStatus(MapLineStatus.RED);
                return;
            }

            line.setMapLineStatus(MapLineStatus.GREEN);
        });
    }

    private void updateTowerInfo() {
        List<Building> allBuldings = findHelper.getAllBuldings(false);

        allBuldings.stream()
                .filter(tower -> tower.getType().equals(BuildingType.GUARDIAN_TOWER))
                .filter(tower -> tower.getLife() < tower.getMaxLife() * DEAD_TOWER_HP_FACTOR)
                .forEach(tower -> {
                    Point2D towerPoint = new Point2D(tower.getX(), tower.getY());

                    deadGuardTowers.put(towerPoint, tower);
                });

        deadGuardTowers.forEach((point2D, tower) -> {
            boolean isEnemy = findHelper.isEnemy(wizard.getFaction(), tower);

            List<LinePosition> linePositions = getLinePositions(tower, 1.5).stream()
                    .filter(linePosition -> linePosition.getMapLine().getEnemy() == isEnemy)
                    .filter(linePosition -> linePosition.getMapLine().getLaneType() != null)
                    .filter(linePosition -> abs(getAngleTo(linePosition.getMapLine(), tower)) < PI / 6)
                    .collect(Collectors.toList());


            MapLine towerLine = linePositions.get(0).getMapLine();

            if (isEnemy)
                towerLine.setDeadEnemyTowerCount(towerLine.getDeadEnemyTowerCount() + 1);
            else
                towerLine.setDeadFriendTowerCount(towerLine.getDeadFriendTowerCount() + 1);
        });
    }

    private void updateWizardPositions() {
        List<Wizard> allWizards = findHelper.getAllWizards(false, false);

        allWizards.forEach(someWizard -> {
            List<LinePosition> linePositions = getLinePositions(someWizard, 1);

            linePositions.forEach(linePosition -> {
                MapLine wizardLine = linePosition.getMapLine();
                Point2D wizardPoint = new Point2D(someWizard.getX(), someWizard.getY());

                if (findHelper.isEnemy(wizard.getFaction(), someWizard))
                    wizardLine.getEnemyWizardPositions().put(wizardPoint, linePosition.getPosition());
                else
                    wizardLine.getFriendWizardPositions().put(wizardPoint, linePosition.getPosition());
            });
        });
    }

    private void updateMinionPositions() {
        clearMinionPosition();
        List<Minion> allMinions = findHelper.getAllMinions(false, false);

        allMinions.forEach(minion -> {
            List<LinePosition> linePositions = getLinePositions(minion, 1);

            linePositions.forEach(linePosition -> {
                MapLine minionLine = linePosition.getMapLine();
                if (findHelper.isEnemy(wizard.getFaction(), minion)) {
                    if (minionLine.getEnemyPosition() > linePosition.getPosition())
                        minionLine.setEnemyPosition(linePosition.getPosition());
                } else {
                    if (minionLine.getFriendPosition() < linePosition.getPosition())
                        minionLine.setFriendPosition(linePosition.getPosition());
                }
            });
        });
    }

    private void clearMinionPosition() {
        mapLines.forEach(mapLine -> {
            mapLine.setEnemyPosition(0);
            mapLine.setEnemyPosition(0);
        });
    }

    public List<LinePosition> getLinePositions(double x, double y, double radiusMultiplier) {
        Point2D searchingPoint = new Point2D(x, y);

        return mapLines.stream()
                .map(line -> {
                    Pair<Double, Double> distance = getDistanceFromLine(searchingPoint, line);

                    double distanceFromLine = distance.getFirst();
                    double linePosition = distance.getSecond();

                    Optional<Pair<Double, LinePosition>> resultLine;
                    if (distanceFromLine <= LINE_RESOLVING_DISTANCE * radiusMultiplier
                            && linePosition >= -LINE_RESOLVING_POSITION * radiusMultiplier && linePosition <= (line.getLineLength() + LINE_RESOLVING_POSITION * radiusMultiplier)) {
                        resultLine = Optional.of(new Pair<>(distanceFromLine, new LinePosition(line, linePosition)));
                    } else
                        resultLine = Optional.empty();

                    return resultLine;
                })
                .filter(Optional::isPresent)
                .map(value -> value.get().getSecond())
                .collect(Collectors.toList());
    }

    public List<LinePosition> getLinePositions(LivingUnit unit, double radiusMultiplier) {
        return getLinePositions(unit.getX(), unit.getY(), radiusMultiplier);
    }

    public List<LinePosition> getLinePositions(Point2D unit, double radiusMultiplier) {
        return getLinePositions(unit.getX(), unit.getY(), radiusMultiplier);
    }

    public Point2D getPointInLine(LinePosition linePosition) {
        return getPointInLine(linePosition.getMapLine(), linePosition.getPosition());
    }

    public Point2D getPointInLine(MapLine line, double nextLineDistance) {
        double newX = line.getStartPoint().getX() + cos(line.getAngle()) * nextLineDistance;
        double newY = line.getStartPoint().getY() + sin(line.getAngle()) * nextLineDistance;

        return new Point2D(newX, newY);
    }

    public Point2D getNearestPointInLine(LivingUnit unit) {
        return getNearestPointInLine(new Point2D(unit.getX(), unit.getY()));
    }

    public Point2D getNearestPointInLine(Point2D point) {
        Point2D nearestPoint = mapPoints.stream()
                .min(Comparator.comparing(pointInLine -> pointInLine.getDistanceTo(point)))
                .get();

        //TODO fix safe calculating
        List<MapLine> nearestLines = mapLines.stream()
                .filter(line -> line.getStartPoint().equals(nearestPoint) || line.getEndPoint().equals(nearestPoint))
                .collect(Collectors.toList());

        MapLine mapLine = nearestLines.stream()
                .min(Comparator.comparing(line -> ((Double) getDistanceFromLine(point, line).getFirst())))
                .get();

        return getPointInLine(mapLine, getDistanceFromLine(point, mapLine).getSecond());
    }

    private Pair<Double, Double> getDistanceFromLine(Point2D point, MapLine line) {
        double angleToPoint = getAngleTo(line, point);
        double distanceToPoint = point.getDistanceTo(line.getStartPoint());

        double distanceFromLine = abs(sin(angleToPoint) * distanceToPoint);
        double linePosition = cos(angleToPoint) * distanceToPoint;

        return new Pair<>(distanceFromLine, linePosition);
    }

    private double getAngleTo(MapLine mapLine, LivingUnit point) {
        return getAngleTo(mapLine, new Point2D(point.getX(), point.getY()));
    }

    private double getAngleTo(MapLine mapLine, Point2D point) {
        Point2D startPoint = mapLine.getStartPoint();

        double absoluteAngleTo = atan2(point.getY() - startPoint.getY(), point.getX() - startPoint.getX());
        double relativeAngleTo = absoluteAngleTo - mapLine.getAngle();

        while (relativeAngleTo > PI) {
            relativeAngleTo -= 2.0D * PI;
        }

        while (relativeAngleTo < -PI) {
            relativeAngleTo += 2.0D * PI;
        }

        return relativeAngleTo;
    }

    public Point2D getLinePointToBaseEnemy(LaneType laneType) {
        MapLine enemyLine = mapLines.stream()
                .filter(mapLine -> mapLine.getLaneType() == laneType)
                .filter(mapLine -> mapLine.getEndPoint().equals(enemyBasePoint))
                .findFirst().get();

        Boolean isInLine = getLinePositions(wizard, 1).stream()
                .anyMatch(linePosition -> linePosition.getMapLine().equals(enemyLine));

        if (isInLine)
            return enemyLine.getEndPoint();
        else
            return enemyLine.getStartPoint();
    }


}

File: src/main/kotlin/MapHelper.kt
import MapHelper.HISTORY_TICK_COUNT
import model.*
import java.lang.StrictMath.*
import java.util.*

data class AttackLine(val friend: MapLine, val enemy: MapLine)

object MapHelper {
    lateinit var world: World
    lateinit var game: Game
    lateinit var wizard: Wizard
    lateinit var findHelper: FindHelper

    var lastCheckPointToMoveTick: Int = 0
    var lastPointToMove: Point2D? = null
    var lastLaneToMove: LaneType? = null

    fun nextTick(world: World, game: Game, wizard: Wizard) {
        this.game = game
        this.world = world
        this.wizard = wizard
        this.findHelper = FindHelper(world, game, wizard)

        updateLineInfo()
    }

    private fun updateLineInfo() {
        clearLinesInfo()

        updateTowerInfo()
        updateWizardPositions()
        updateMinionPositions()
    }

    private fun clearLinesInfo() {
        mapLines.forEach { mapLine ->
            mapLine.enemyWizardPositions.clear()
            mapLine.friendWizardPositions.clear()

            mapLine.historyEnemyWizardPositions.filterByTick(world.tickIndex)
            mapLine.historyFriendWizardPositions.filterByTick(world.tickIndex)

            if (mapLine.enemy) {
                mapLine.enemyPosition = mapLine.lineLength
                mapLine.friendPosition = null
            } else {
                mapLine.enemyPosition = null
                mapLine.friendPosition = 0.0
            }

            mapLine.deadEnemyTowerCount = 0
            mapLine.deadFriendTowerCount = 0
        }
    }

    private fun updateTowerInfo() {
        val allBuldings = findHelper.getAllBuldings(false)

        allBuldings
                .filter { tower -> tower.type == BuildingType.GUARDIAN_TOWER }
                .filter { tower -> tower.life < tower.maxLife * DEAD_TOWER_HP_FACTOR }
                .forEach { tower ->
                    deadGuardTowers.put(tower.id, tower)
                }

        deadGuardTowers.forEach { value ->
            val (poinn2D, tower) = value
            val isEnemy = findHelper.isEnemy(wizard.faction, tower)

            val linePosition = getLinePositions(tower, 5.0)
                    .filter { linePosition -> linePosition.mapLine.enemy == isEnemy }
                    .filter { linePosition -> linePosition.mapLine.laneType != null }
                    .filter { linePosition -> linePosition.position >= 0 && linePosition.position <= linePosition.mapLine.lineLength }
                    .filter { linePosition -> abs(linePosition.mapLine.getAngleTo(tower)) < PI / 5 }
                    .minBy { getDistanceFromLine(tower.toPoint(), it.mapLine).first }

            if (linePosition == null)
                throw RuntimeException("invalid tower")

            val towerLine = linePosition.mapLine

            if (isEnemy)
                towerLine.deadEnemyTowerCount += 1
            else
                towerLine.deadFriendTowerCount += 1

            if (towerLine.deadFriendTowerCount > 2 || towerLine.deadEnemyTowerCount > 2)
                throw RuntimeException("omg.. ")
        }
    }

    private fun updateWizardPositions() {
        val allWizards = (findHelper.getAllWizards(false, false) + wizard).toSet()

        allWizards.forEach { someWizard ->
            val linePositions = getLinePositions(someWizard, 1.0)

            linePositions.forEach { linePosition ->
                val wizardLine = linePosition.mapLine
                val wizardId = someWizard.id

                if (findHelper.isEnemy(wizard.faction, someWizard)) {
                    wizardLine.enemyWizardPositions.put(wizardId, linePosition.position)
                    wizardLine.historyEnemyWizardPositions.put(wizardId, HistoryValue(world.tickIndex, linePosition.position))
                } else {
                    wizardLine.friendWizardPositions.put(wizardId, linePosition.position)
                    wizardLine.historyFriendWizardPositions.put(wizardId, HistoryValue(world.tickIndex, linePosition.position))
                }
            }
        }
    }

    private fun updateMinionPositions() {
        clearMinionPosition()
        val allMinions = findHelper.getAllMinions(false, false)

        allMinions.forEach { minion ->
            val linePositions = getLinePositions(minion, 2.0)

            linePositions.forEach { linePosition ->
                val minionLine = linePosition.mapLine
                if (findHelper.isEnemy(wizard.faction, minion)) {
                    if (minionLine.enemyPosition == null || (minionLine.enemyPosition ?: minionLine.lineLength) < linePosition.position)
                        minionLine.enemyPosition = linePosition.position
                } else {
                    if (minionLine.friendPosition == null || (minionLine.friendPosition ?: 0.0) < linePosition.position)
                        minionLine.friendPosition = linePosition.position
                }
            }
        }
    }

    private fun clearMinionPosition() {
        mapLines.forEach { mapLine ->
            if (mapLine.enemy) {
                mapLine.enemyPosition = mapLine.lineLength
                mapLine.friendPosition = null
            } else {
                mapLine.enemyPosition = null
                mapLine.friendPosition = 0.0
            }
        }
    }

    fun getLinePositions(x: Double, y: Double, radiusMultiplier: Double): List<LinePosition> {
        val searchingPoint = Point2D(x, y)

        return mapLines
                .map { line ->
                    val distance = getDistanceFromLine(searchingPoint, line)

                    val distanceFromLine = distance.first
                    val linePosition = distance.second

                    val resultLine: Optional<Pair<Double, LinePosition>>
                    if (distanceFromLine <= LINE_RESOLVING_DISTANCE * radiusMultiplier
                            && linePosition >= -LINE_RESOLVING_POSITION * radiusMultiplier && linePosition <= line.lineLength + LINE_RESOLVING_POSITION * radiusMultiplier) {
                        resultLine = Optional.of(Pair(distanceFromLine, LinePosition(line, linePosition)))
                    } else
                        resultLine = Optional.empty<Pair<Double, LinePosition>>()

                    resultLine
                }
                .filter { it.isPresent }
                .map { value -> value.get().second }
    }

    fun getLinePositions(unit: LivingUnit, radiusMultiplier: Double): List<LinePosition> {
        return getLinePositions(unit.x, unit.y, radiusMultiplier)
    }

    fun getLinePositions(unit: Point2D, radiusMultiplier: Double): List<LinePosition> {
        return getLinePositions(unit.x, unit.y, radiusMultiplier)
    }

    fun getPointInLine(linePosition: LinePosition): Point2D {
        return getPointInLine(linePosition.mapLine, linePosition.position)
    }

    fun getPointInLine(line: MapLine, nextLineDistance: Double): Point2D {
        val newX = line.startPoint.x + cos(line.angle) * nextLineDistance
        val newY = line.startPoint.y + sin(line.angle) * nextLineDistance

        return Point2D(newX, newY)
    }

    fun getNearestPointInLine(unit: LivingUnit): Point2D {
        return getNearestPointInLine(Point2D(unit.x, unit.y))
    }

    fun getNearestPointInLine(point: Point2D): Point2D {
        val nearestPoint = mapPoints
                .minBy { pointInLine -> pointInLine.getDistanceTo(point) }

        //TODO fix safe calculating
        val nearestLines = mapLines
                .filter { line -> line.startPoint == nearestPoint || line.endPoint == nearestPoint }

        val mapLine = nearestLines
                .minBy { line -> getDistanceFromLine(point, line).first }

        return getPointInLine(mapLine!!, getDistanceFromLine(point, mapLine).second)
    }

    private fun getDistanceFromLine(point: Point2D, line: MapLine): Pair<Double, Double> {
        val angleToPoint = line.getAngleTo(point)
        val distanceToPoint = point.getDistanceTo(line.startPoint)

        val distanceFromLine = abs(sin(angleToPoint) * distanceToPoint)
        val linePosition = cos(angleToPoint) * distanceToPoint

        return Pair(distanceFromLine, linePosition)
    }

    private fun MapLine.getAngleTo(point: LivingUnit): Double {
        return this.getAngleTo(Point2D(point.x, point.y))
    }

    private fun MapLine.getAngleTo(point: Point2D): Double {
        val startPoint = this.startPoint

        val absoluteAngleTo = atan2(point.y - startPoint.y, point.x - startPoint.x)
        var relativeAngleTo = absoluteAngleTo - this.angle

        while (relativeAngleTo > PI) {
            relativeAngleTo -= 2.0 * PI
        }

        while (relativeAngleTo < -PI) {
            relativeAngleTo += 2.0 * PI
        }

        return relativeAngleTo
    }

    fun getLinePointToBaseEnemy(laneType: LaneType): Point2D {
        if (world.tickIndex - lastCheckPointToMoveTick < CHANGE_POINT_TO_MOVE_MIN_TICK_DIFF
                && lastPointToMove != null && lastLaneToMove == laneType)
            return lastPointToMove!!

        val lane = attackLines[laneType]!!

        lastCheckPointToMoveTick = world.tickIndex
        lastLaneToMove = laneType

        lastPointToMove = if (lane.enemy.friendPosition != null)
            getPointInLine(lane.enemy, lane.enemy.friendPosition!! + MOVE_FORWARD)
        else if (lane.friend.friendPosition != 0.0)
            if (lane.friend.friendPosition!! < lane.friend.lineLength * LINE_BACK_FACTOR)
                getPointInLine(lane.friend, lane.friend.enemyPosition ?: lane.friend.friendPosition!!)
            else
                getPointInLine(lane.friend, max(lane.friend.friendPosition ?: 0.0, lane.friend.enemyPosition ?: 0.0) + MOVE_FORWARD)
        else {
            getPointInLine(lane.friend, lane.friend.lineLength - START_POINT_POSITON)
        }

        return lastPointToMove!!
    }


    val mapLines: MutableList<MapLine> = mutableListOf()

    val attackLines: MutableMap<LaneType, AttackLine> = mutableMapOf()

    val mapPoints: MutableList<Point2D> = mutableListOf()

    var mapSize = 4000.0

    val friendBasePoint = Point2D(150.0, mapSize - 150.0)
    var topPoint = Point2D(300.0, 300.0)
    var middlePoint = Point2D(2000.0, 2000.0)
    var bottomPoint = Point2D(mapSize - 300.0, mapSize - 300.0)
    var enemyBasePoint = Point2D(mapSize - 100.0, 100.0)

    var topFriendLine = MapLine(friendBasePoint, topPoint, LaneType.TOP, false)
    var topEnemyLine = MapLine(topPoint, enemyBasePoint, LaneType.TOP, true)

    var middleFriendLine = MapLine(friendBasePoint, middlePoint, LaneType.MIDDLE, false)
    var middleEnemyLine = MapLine(middlePoint, enemyBasePoint, LaneType.MIDDLE, true)

    var bottomFriendLine = MapLine(friendBasePoint, bottomPoint, LaneType.BOTTOM, false)
    var bottomEnemyLine = MapLine(bottomPoint, enemyBasePoint, LaneType.BOTTOM, true)

    var artifactTopLine = MapLine(topPoint, middlePoint, null)
    var artifactBottomLine = MapLine(middlePoint, bottomPoint, null)

    init {
        topFriendLine.startLines.addAll(Arrays.asList(middleFriendLine, bottomFriendLine))
        topFriendLine.endLines.addAll(Arrays.asList(topEnemyLine, artifactTopLine))

        topEnemyLine.startLines.addAll(Arrays.asList(topFriendLine, artifactTopLine))
        topEnemyLine.endLines.addAll(Arrays.asList(middleEnemyLine, bottomEnemyLine))

        middleFriendLine.startLines.addAll(Arrays.asList(topFriendLine, bottomFriendLine))
        middleFriendLine.endLines.addAll(Arrays.asList(topEnemyLine, artifactBottomLine, artifactTopLine))

        middleEnemyLine.startLines.addAll(Arrays.asList(topFriendLine, artifactBottomLine, artifactTopLine))
        middleEnemyLine.endLines.addAll(Arrays.asList(topEnemyLine, bottomEnemyLine))

        bottomFriendLine.startLines.addAll(Arrays.asList(topFriendLine, middleFriendLine))
        bottomFriendLine.endLines.addAll(Arrays.asList(artifactBottomLine, bottomEnemyLine))

        bottomEnemyLine.startLines.addAll(Arrays.asList(artifactBottomLine, bottomFriendLine))
        bottomEnemyLine.endLines.addAll(Arrays.asList(topEnemyLine, middleEnemyLine))

        artifactTopLine.startLines.addAll(Arrays.asList(topFriendLine, topEnemyLine))
        artifactTopLine.endLines.addAll(Arrays.asList(middleFriendLine, middleEnemyLine, artifactBottomLine))

        artifactBottomLine.startLines.addAll(Arrays.asList(middleFriendLine, middleEnemyLine, artifactTopLine))
        artifactBottomLine.endLines.addAll(Arrays.asList(bottomFriendLine, bottomEnemyLine))

        mapLines.addAll(Arrays.asList(
                topFriendLine, topEnemyLine, middleFriendLine, middleEnemyLine, bottomFriendLine, bottomEnemyLine, artifactTopLine, artifactBottomLine
        ))

        mapPoints.addAll(Arrays.asList(
                friendBasePoint, topPoint, middlePoint, bottomPoint, enemyBasePoint
        ))

        attackLines.put(LaneType.TOP, AttackLine(topFriendLine, topEnemyLine))
        attackLines.put(LaneType.MIDDLE, AttackLine(middleFriendLine, middleEnemyLine))
        attackLines.put(LaneType.BOTTOM, AttackLine(bottomFriendLine, bottomEnemyLine))
    }

    val LINE_RESOLVING_POSITION = 250.0
    val LINE_RESOLVING_DISTANCE = 250.0

    val DEAD_TOWER_HP_FACTOR = 0.1

    val START_POINT_POSITON = 300.0

    val MOVE_FORWARD = 200.0

    val LINE_BACK_FACTOR: Double = 0.5

    val CHANGE_POINT_TO_MOVE_MIN_TICK_DIFF: Int = 50

    val HISTORY_TICK_COUNT: Int = 2000

    var deadGuardTowers: MutableMap<Long, Building> = HashMap()
}

fun MutableMap<Long, HistoryValue>.filterByTick(currentTick: Int) {
    val iterator = this.iterator()
    while(iterator.hasNext()){
        val currentValue = iterator.next()

        if (currentTick - currentValue.value.tick > HISTORY_TICK_COUNT)
            iterator.remove()
    }
}

fun Map<Long, HistoryValue>.toData(): Map<Long, Double> {
    return this.mapValues { it.value.value }
}
